xarray/backends/api.py

```diff
@@ -980,32 +1020,100 @@ def multi_file_closer():
     # read global attributes from the attrs_file or from the first dataset
     if attrs_file is not None:
         if isinstance(attrs_file, os.PathLike):
-            attrs_file = os.fspath(attrs_file)
+            attrs_file = cast(str, os.fspath(attrs_file))
         combined.attrs = datasets[paths.index(attrs_file)].attrs
 
     return combined
 
 
-WRITEABLE_STORES: Dict[str, Callable] = {
+WRITEABLE_STORES: dict[T_NETCDFENGINE, Callable] = {
     "netcdf4": backends.NetCDF4DataStore.open,
     "scipy": backends.ScipyDataStore,
     "h5netcdf": backends.H5NetCDFStore.open,
 }
 
 
+@overload
+def to_netcdf(
+    dataset: Dataset,
+    path_or_file: str | os.PathLike | None,
+    mode: Literal["w", "a"],
+    format: T_NETCDFTYPES | None,
+    group: str | None,
+    engine: T_NETCDFENGINE | None,
+    encoding: Mapping[Hashable, Mapping[str, Any]] | None,
+    unlimited_dims: Iterable[Hashable] | None,
+    compute: bool,
+    multifile: Literal[True],
+    invalid_netcdf: bool,
+) -> tuple[ArrayWriter, AbstractDataStore]:
+    ...
+
+
+@overload
 def to_netcdf(
     dataset: Dataset,
-    path_or_file=None,
-    mode: str = "w",
-    format: str = None,
-    group: str = None,
-    engine: str = None,
-    encoding: Mapping = None,
-    unlimited_dims: Iterable[Hashable] = None,
+    path_or_file: None,
+    mode: Literal["w", "a"],
+    format: T_NETCDFTYPES | None,
+    group: str | None,
+    engine: T_NETCDFENGINE | None,
+    encoding: Mapping[Hashable, Mapping[str, Any]] | None,
+    unlimited_dims: Iterable[Hashable] | None,
+    compute: bool,
+    multifile: Literal[False],
+    invalid_netcdf: bool,
+) -> bytes:
+    ...
+
+
+@overload
+def to_netcdf(
+    dataset: Dataset,
+    path_or_file: str | os.PathLike,
+    mode: Literal["w", "a"],
+    format: T_NETCDFTYPES | None,
+    group: str | None,
+    engine: T_NETCDFENGINE | None,
+    encoding: Mapping[Hashable, Mapping[str, Any]] | None,
+    unlimited_dims: Iterable[Hashable] | None,
+    compute: Literal[False],
+    multifile: Literal[False],
+    invalid_netcdf: bool,
+) -> Delayed:
+    ...
+
+
+@overload
+def to_netcdf(
+    dataset: Dataset,
+    path_or_file: str | os.PathLike,
+    mode: Literal["w", "a"],
+    format: T_NETCDFTYPES | None,
+    group: str | None,
+    engine: T_NETCDFENGINE | None,
+    encoding: Mapping[Hashable, Mapping[str, Any]] | None,
+    unlimited_dims: Iterable[Hashable] | None,
+    compute: Literal[True],
+    multifile: Literal[False],
+    invalid_netcdf: bool,
+) -> None:
+    ...
+
+
+def to_netcdf(
```

Comment:
To the extent we can add a phrase on each overload for what each one is, that may make it easier to grok, particularly for these quite mega-methods. Though tbf looking at the return type gives most of the story, so only if you think it'd be helpful.